apps/src/lib/levelbuilder/code-docs-editor/ProgrammingExpressionEditor.jsx

```diff
@@ -0,0 +1,132 @@
+import PropTypes from 'prop-types';
+import React, {useState} from 'react';
+import TextareaWithMarkdownPreview from '@cdo/apps/lib/levelbuilder/TextareaWithMarkdownPreview';
+import SaveBar from '@cdo/apps/lib/levelbuilder/SaveBar';
+import {navigateToHref} from '@cdo/apps/utils';
+import $ from 'jquery';
+import color from '@cdo/apps/util/color';
+
+export default function ProgrammingExpressionEditor({
+  initialProgrammingExpression
+}) {
+  const [programmingExpression, setProgrammingExpression] = useState({
+    ...initialProgrammingExpression,
+    id: null,
+    key: null
+  });
+  const [isSaving, setIsSaving] = useState(false);
+  const [lastUpdated, setLastUpdated] = useState(
+    initialProgrammingExpression.lastUpdated
+  );
+  const [error, setError] = useState(null);
+
+  const save = () => {
+    if (isSaving) {
+      return;
+    }
+    setIsSaving(true);
+    fetch(`/programming_expressions/${initialProgrammingExpression.id}`, {
+      method: 'PUT',
+      headers: {
+        'content-type': 'application/json',
+        'X-CSRF-Token': $('meta[name="csrf-token"]').attr('content')
+      },
+      body: JSON.stringify(programmingExpression)
+    })
+      .then(response => {
+        if (response.ok) {
+          setIsSaving(false);
+          setLastUpdated(Date.now());
+        } else {
+          setIsSaving(false);
+          setError(response.statusText);
+        }
+      })
+      .catch(error => {
+        setIsSaving(false);
+        setError(error.responseText);
+      });
+  };
+
+  return (
+    <div>
+      <h1>{`Editing ${initialProgrammingExpression.key}`}</h1>
+      <h2>
+        This feature is in development. Please continue to use curriculum
+        builder to edit code documentation.
+      </h2>
+      <label>
+        Display Name
+        <input
+          value={programmingExpression.name}
+          onChange={e =>
+            setProgrammingExpression({
+              ...programmingExpression,
+              name: e.target.value
+            })
+          }
```

Comment:
since we'll be adding a bunch of state to programmingExpression, we should wrap this in another function to simplify these `onChange` handlers

```javascript
// caveat: this is kind of pseudocode
const onChangeProgrammingExpression = (key, value) => {
  setProgrammingExpression({...programmingExpression, [key]: value})
}

return <input onChange={e => onChangeProgrammingExpression('name', e.target.value)} />
```

Reply:
That makes sense but for my understanding, why not use a reducer here to manage this? Maybe that's what you were talking about with the custom hooks?

Reply:
yeah, that's what i was thinking with a custom hook. i don't think we need to use [useReducer](https://reactjs.org/docs/hooks-reference.html#usereducer) because there aren't distinct actions or complex state management, so a hook would probably be more appropriate. it would end up being pretty simple, but would clean up this component a bit because it could encapsulate a few things:

1. the logic to destructure/exclude certain data from `initialProgrammingExpression`
2. put all the appropriate data in state with useState
3. manage updating state for a particular key

i imagine the hook could look something like

```javascript
function useProgrammingExpression(initialProgrammingExpression) {
  // remove any data from initialProgrammingExpression that we don't want in state,
  // or do that before calling this method
  const [programmingExpression, setProgrammingExpression] = useState(initialProgrammingExpression);

  function updateProgrammingExpression(key, value) {
    setProgrammingExpression({...programmingExpression, [key]: value});
  }

  return [programmingExpression, updateProgrammingExpression];
}
```

Reply:
oh that is a lot cleaner! I'll give this a try